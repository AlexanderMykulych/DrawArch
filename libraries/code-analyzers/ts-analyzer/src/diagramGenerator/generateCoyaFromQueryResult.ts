import type { QueryResult } from 'neo4j-driver';
import { applyPositioning } from 'coya-core'
import { progressSync } from '../progress/progress';

const getBlockId = (id: string) => `b_${id}`;
const getArrowId = (id: string) => `a_${id}`;

function _generateCoyaFromGraphResult(result: QueryResult) {
  const blocksMap = new Map<string, any>();
  const arrowsMap = new Map<string, any>();
  result
    .records
    .forEach((item) => {
      const graph = item.get('graph');
      graph.relationships.forEach((r) => {
        arrowsMap.set(getArrowId(r.identity), {
          from: getBlockId(r.start.toString()),
          to: getBlockId(r.end.toString()),
          type: 'line',
          label: '',
        });
      });
      graph.nodes.forEach(n => blocksMap.set(getBlockId(n.identity), n));
    });
  // .forEach((item) => {
  //     const from = item.get('from');
  //     const to = item.get('to');
  //     const r = item.get('r');
  //     blocksMap.set(getBlockId(from.identity), from.properties.name);
  //     blocksMap.set(getBlockId(to.identity), to.properties.name);
  //     arrowsMap.set(getArrowId(r.identity), {
  //         from: getBlockId(r.start.toString()),
  //         to: getBlockId(r.end.toString()),
  //         type: 'line',
  //         label: '',
  //     });
  // });

  const blocksEntries = [...blocksMap.entries()];
  const blocks = Object.fromEntries(blocksEntries.map(([key, node]) => [key, node.properties.id]));
  const arrows = Object.fromEntries(arrowsMap.entries());
  const diagram = {
    name: 'coya-ts-analyzer',
    blocks: {
      ...blocks,
      ...arrows,
    },
    phases: [],
    style: {
      blocks: {
        _: {
          css: {
            text: {
              fontSize: 'auto',
            },
          },
        },
        _folder: {
          css: {
            fillStyle: 'solid',
            color: '#ffffffff',
            fill: '#f67c01ff',
          },
        },
        _file: {
          css: {
            fillStyle: 'solid',
            color: '#ffffffff',
            fill: '#008781ff',
          },
        },
        _func: {
          css: {
            fillStyle: 'solid',
            color: '#ffffffff',
            fill: '#2097f3ff',
          },
        },
        ...Object.fromEntries(
          blocksEntries.map(([key, val]) => [key, {
            css: {
              get: val.properties.entityType === 'file'
                ? '_file'
                : val.properties.entityType === 'folder'
                  ? '_folder'
                  : '_func',
            },
          }]),
        ),
      },
    },
  }

  applyPositioning(diagram, 'Dagre')

  return diagram;
}

export const generateCoyaFromGraphResult = progressSync('neo4j data to coya', _generateCoyaFromGraphResult)