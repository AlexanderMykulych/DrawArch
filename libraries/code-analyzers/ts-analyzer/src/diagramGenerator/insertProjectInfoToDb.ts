import { analyze } from "../analysis/analyze";
import { useNeo4j } from "../neo4j/useNeo4j";


export async function insertProjectInfoToDb(path: string) {
  const codeInfos = await analyze(path);

  const db = useNeo4j();
  await db.init();
  await db.insertCodeInfos(codeInfos);

  return {
    db,
    codeInfos,
  }
}
