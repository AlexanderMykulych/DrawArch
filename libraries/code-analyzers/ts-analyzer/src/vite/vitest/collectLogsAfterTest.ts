import { ActionEntity, ActionEntityType, CodeInfo, CodeInfoType, Relationship, RelationType } from "../../analysis/types"
import { useNeo4j } from "../../neo4j/useNeo4j"
import type { LogItem, StartLogItem } from "../shim"


export async function collectLogsAfterTest(insertToDb: true): Promise<void>
export async function collectLogsAfterTest(): Promise<CodeInfo[]>
export async function collectLogsAfterTest(insertToDb?: true): Promise<CodeInfo[] | void> {
  if (!globalThis.__getLogs) {
    return
  }

  const logs = globalThis.__getLogs()

  const codeInfos = logsToCodeInfo(logs)

  if (insertToDb) {
    const db = useNeo4j()
    await db.insertCodeInfos(codeInfos);
  } else {
    return codeInfos
  }
}


export function logsToCodeInfo(logs: LogItem[]): CodeInfo[] {
  return logs
    .flatMap<CodeInfo>((item, index, items) => {
      const result = []
      if (item.type === 'end') {
        const callEntity: ActionEntity = {
          type: CodeInfoType.Entity,
          entityType: ActionEntityType.Call,
          id: getCallActionId(item, index),
          executionTime: item.execTime,
        }
        const callRelation: Relationship = {
          type: CodeInfoType.Relationship,
          relationType: RelationType.CallEntity,
          from: callEntity.id,
          to: item.id,
        }

        result.push(callEntity, callRelation)

        const [parentStartItem, parentIndex] = findParentStartLogFromIndex(items, index, item.id)

        if (parentStartItem) {
          const parentRelation: Relationship = {
            type: CodeInfoType.Relationship,
            relationType: RelationType.Parent,
            from: getCallActionId(parentStartItem, parentIndex),
            to: callEntity.id,
          }
          result.push(parentRelation)
        }
      }
      return result
    })
}

export function findStartLogFromIndex(items: LogItem[], index: number, id: string): StartLogItem {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id === id) {
      return startItem
    }
  }
  throw 'Can not find start element for: ' + id
}

export function findParentStartLogFromIndex(items: LogItem[], index: number, id: string): [StartLogItem, number] | [null, null] {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id !== id) {
      return [startItem, i]
    }
  }

  return [null, null]
}

export function getCallActionId(item: LogItem, index: number): string {
  // return `call:${item.id}`
  return `call[${index}]:${item.id}`
}