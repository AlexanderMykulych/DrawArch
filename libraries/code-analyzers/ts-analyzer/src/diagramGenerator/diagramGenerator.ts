import { generateCoyaFromGraphResult } from "./generateCoyaFromQueryResult";
import { insertProjectInfoToDb } from "./insertProjectInfoToDb";

export async function diagramGenerator(path: string) {
  const { db } = await insertProjectInfoToDb(path)

  const graph = await db.read('match p = (n1)-[*..]->() return apoc.agg.graph(p) as graph')

  if (!graph) {
    throw 'graph is empty'
  }

  const coya = generateCoyaFromGraphResult(graph)

  return {
    db,
    coya,
  }
}
