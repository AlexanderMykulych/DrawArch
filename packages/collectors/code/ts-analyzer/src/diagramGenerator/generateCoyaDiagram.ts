import { generateCoyaFromGraphResult } from "./generateCoyaFromQueryResult";
import { insertProjectInfoToDb } from "./insertProjectInfoToDb";

export async function generateCoyaDiagram(path: string) {
  const { db } = await insertProjectInfoToDb(path)

  const graph = await db.read(`
  match p = (n1)-[*..{relationType: "parent"}]->(n2)
  where
    not n1.id contains 'node_modules/'
    and
    not n2.id contains 'node_modules/'
  return apoc.agg.graph(p) as graph
`)
  // const graph = await db.read('match p = (n1)-[*0..]->() return apoc.agg.graph(p) as graph')

  if (!graph) {
    throw 'graph is empty'
  }

  const coya = generateCoyaFromGraphResult(graph)

  return {
    db,
    coya,
  }
}
