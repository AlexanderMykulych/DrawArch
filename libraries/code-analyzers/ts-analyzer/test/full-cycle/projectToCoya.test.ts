import path from 'path'
import { test, expect } from 'vitest'
import { diagramGenerator } from '../../src/diagramGenerator/diagramGenerator'


test.each([
  {
    projectPath: '/cases/01_simple',
  },
  {
    projectPath: '/cases/02_function',
  },
  {
    projectPath: '/cases/03_function_relation',
  },
  {
    projectPath: '/cases/04_vue',
  },
])('should return project coya diagram: $projectPath', async ({ projectPath }) => {
  const fullProjectPath = path.join(__dirname, projectPath)

  const { coya } = await diagramGenerator(fullProjectPath)

  expect(coya).toMatchSnapshot()
})
