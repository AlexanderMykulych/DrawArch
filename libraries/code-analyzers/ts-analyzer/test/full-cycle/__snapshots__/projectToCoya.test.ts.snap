// Vitest Snapshot v1

exports[`should return project coya diagram: /cases/01_simple 1`] = `
{
  "blocks": {
    "a_7": {
      "from": "b_2",
      "label": "",
      "to": "b_1",
      "type": "line",
    },
    "a_8": {
      "from": "b_3",
      "label": "",
      "to": "b_2",
      "type": "line",
    },
    "b_1": undefined,
    "b_2": undefined,
    "b_3": undefined,
  },
  "name": "coya-ts-analyzer",
  "phases": [],
  "style": {
    "blocks": {
      "_": {
        "css": {
          "text": {
            "fontSize": "auto",
          },
        },
      },
      "_file": {
        "css": {
          "color": "#ffffffff",
          "fill": "#008781ff",
          "fillStyle": "solid",
        },
      },
      "_folder": {
        "css": {
          "color": "#ffffffff",
          "fill": "#f67c01ff",
          "fillStyle": "solid",
        },
      },
      "_func": {
        "css": {
          "color": "#ffffffff",
          "fill": "#2097f3ff",
          "fillStyle": "solid",
        },
      },
      "b_1": {
        "css": {
          "get": "_func",
        },
      },
      "b_2": {
        "css": {
          "get": "_func",
        },
      },
      "b_3": {
        "css": {
          "get": "_func",
        },
      },
    },
    "layout": {
      "type": "Dagre",
    },
  },
}
`;

exports[`should return project coya diagram: /cases/02_function 1`] = `
{
  "blocks": {
    "a_10": {
      "from": "b_47",
      "label": "",
      "to": "b_46",
      "type": "line",
    },
    "a_9": {
      "from": "b_47",
      "label": "",
      "to": "b_45",
      "type": "line",
    },
    "b_45": undefined,
    "b_46": undefined,
    "b_47": undefined,
  },
  "name": "coya-ts-analyzer",
  "phases": [],
  "style": {
    "blocks": {
      "_": {
        "css": {
          "text": {
            "fontSize": "auto",
          },
        },
      },
      "_file": {
        "css": {
          "color": "#ffffffff",
          "fill": "#008781ff",
          "fillStyle": "solid",
        },
      },
      "_folder": {
        "css": {
          "color": "#ffffffff",
          "fill": "#f67c01ff",
          "fillStyle": "solid",
        },
      },
      "_func": {
        "css": {
          "color": "#ffffffff",
          "fill": "#2097f3ff",
          "fillStyle": "solid",
        },
      },
      "b_45": {
        "css": {
          "get": "_func",
        },
      },
      "b_46": {
        "css": {
          "get": "_func",
        },
      },
      "b_47": {
        "css": {
          "get": "_func",
        },
      },
    },
    "layout": {
      "type": "Dagre",
    },
  },
}
`;

exports[`should return project coya diagram: /cases/03_function_relation 1`] = `
{
  "blocks": {
    "a_11": {
      "from": "b_9",
      "label": "",
      "to": "b_4",
      "type": "line",
    },
    "a_12": {
      "from": "b_9",
      "label": "",
      "to": "b_5",
      "type": "line",
    },
    "a_13": {
      "from": "b_48",
      "label": "",
      "to": "b_9",
      "type": "line",
    },
    "a_14": {
      "from": "b_6",
      "label": "",
      "to": "b_49",
      "type": "line",
    },
    "a_15": {
      "from": "b_7",
      "label": "",
      "to": "b_49",
      "type": "line",
    },
    "a_16": {
      "from": "b_8",
      "label": "",
      "to": "b_49",
      "type": "line",
    },
    "a_17": {
      "from": "b_8",
      "label": "",
      "to": "b_7",
      "type": "line",
    },
    "a_18": {
      "from": "b_48",
      "label": "",
      "to": "b_6",
      "type": "line",
    },
    "a_19": {
      "from": "b_9",
      "label": "",
      "to": "b_49",
      "type": "line",
    },
    "a_20": {
      "from": "b_48",
      "label": "",
      "to": "b_7",
      "type": "line",
    },
    "a_21": {
      "from": "b_48",
      "label": "",
      "to": "b_8",
      "type": "line",
    },
    "a_22": {
      "from": "b_9",
      "label": "",
      "to": "b_7",
      "type": "line",
    },
    "b_4": undefined,
    "b_48": undefined,
    "b_49": undefined,
    "b_5": undefined,
    "b_6": undefined,
    "b_7": undefined,
    "b_8": undefined,
    "b_9": undefined,
  },
  "name": "coya-ts-analyzer",
  "phases": [],
  "style": {
    "blocks": {
      "_": {
        "css": {
          "text": {
            "fontSize": "auto",
          },
        },
      },
      "_file": {
        "css": {
          "color": "#ffffffff",
          "fill": "#008781ff",
          "fillStyle": "solid",
        },
      },
      "_folder": {
        "css": {
          "color": "#ffffffff",
          "fill": "#f67c01ff",
          "fillStyle": "solid",
        },
      },
      "_func": {
        "css": {
          "color": "#ffffffff",
          "fill": "#2097f3ff",
          "fillStyle": "solid",
        },
      },
      "b_4": {
        "css": {
          "get": "_func",
        },
      },
      "b_48": {
        "css": {
          "get": "_func",
        },
      },
      "b_49": {
        "css": {
          "get": "_func",
        },
      },
      "b_5": {
        "css": {
          "get": "_func",
        },
      },
      "b_6": {
        "css": {
          "get": "_func",
        },
      },
      "b_7": {
        "css": {
          "get": "_func",
        },
      },
      "b_8": {
        "css": {
          "get": "_func",
        },
      },
      "b_9": {
        "css": {
          "get": "_func",
        },
      },
    },
    "layout": {
      "type": "Dagre",
    },
  },
}
`;

exports[`should return project coya diagram: /cases/04_vue 1`] = `
{
  "blocks": {
    "a_0": {
      "from": "b_16",
      "label": "",
      "to": "b_20",
      "type": "line",
    },
    "a_1": {
      "from": "b_16",
      "label": "",
      "to": "b_17",
      "type": "line",
    },
    "a_2": {
      "from": "b_25",
      "label": "",
      "to": "b_30",
      "type": "line",
    },
    "a_23": {
      "from": "b_25",
      "label": "",
      "to": "b_31",
      "type": "line",
    },
    "a_24": {
      "from": "b_25",
      "label": "",
      "to": "b_32",
      "type": "line",
    },
    "a_25": {
      "from": "b_26",
      "label": "",
      "to": "b_33",
      "type": "line",
    },
    "a_26": {
      "from": "b_26",
      "label": "",
      "to": "b_27",
      "type": "line",
    },
    "a_27": {
      "from": "b_26",
      "label": "",
      "to": "b_0",
      "type": "line",
    },
    "a_28": {
      "from": "b_26",
      "label": "",
      "to": "b_28",
      "type": "line",
    },
    "a_29": {
      "from": "b_16",
      "label": "",
      "to": "b_25",
      "type": "line",
    },
    "a_30": {
      "from": "b_16",
      "label": "",
      "to": "b_26",
      "type": "line",
    },
    "a_31": {
      "from": "b_17",
      "label": "",
      "to": "b_27",
      "type": "line",
    },
    "a_32": {
      "from": "b_18",
      "label": "",
      "to": "b_0",
      "type": "line",
    },
    "a_33": {
      "from": "b_18",
      "label": "",
      "to": "b_28",
      "type": "line",
    },
    "a_34": {
      "from": "b_19",
      "label": "",
      "to": "b_16",
      "type": "line",
    },
    "a_35": {
      "from": "b_19",
      "label": "",
      "to": "b_29",
      "type": "line",
    },
    "b_0": undefined,
    "b_16": undefined,
    "b_17": undefined,
    "b_18": undefined,
    "b_19": undefined,
    "b_20": undefined,
    "b_25": undefined,
    "b_26": undefined,
    "b_27": undefined,
    "b_28": undefined,
    "b_29": undefined,
    "b_30": undefined,
    "b_31": undefined,
    "b_32": undefined,
    "b_33": undefined,
  },
  "name": "coya-ts-analyzer",
  "phases": [],
  "style": {
    "blocks": {
      "_": {
        "css": {
          "text": {
            "fontSize": "auto",
          },
        },
      },
      "_file": {
        "css": {
          "color": "#ffffffff",
          "fill": "#008781ff",
          "fillStyle": "solid",
        },
      },
      "_folder": {
        "css": {
          "color": "#ffffffff",
          "fill": "#f67c01ff",
          "fillStyle": "solid",
        },
      },
      "_func": {
        "css": {
          "color": "#ffffffff",
          "fill": "#2097f3ff",
          "fillStyle": "solid",
        },
      },
      "b_0": {
        "css": {
          "get": "_func",
        },
      },
      "b_16": {
        "css": {
          "get": "_func",
        },
      },
      "b_17": {
        "css": {
          "get": "_func",
        },
      },
      "b_18": {
        "css": {
          "get": "_func",
        },
      },
      "b_19": {
        "css": {
          "get": "_func",
        },
      },
      "b_20": {
        "css": {
          "get": "_func",
        },
      },
      "b_25": {
        "css": {
          "get": "_func",
        },
      },
      "b_26": {
        "css": {
          "get": "_func",
        },
      },
      "b_27": {
        "css": {
          "get": "_func",
        },
      },
      "b_28": {
        "css": {
          "get": "_func",
        },
      },
      "b_29": {
        "css": {
          "get": "_func",
        },
      },
      "b_30": {
        "css": {
          "get": "_func",
        },
      },
      "b_31": {
        "css": {
          "get": "_func",
        },
      },
      "b_32": {
        "css": {
          "get": "_func",
        },
      },
      "b_33": {
        "css": {
          "get": "_func",
        },
      },
    },
    "layout": {
      "type": "Dagre",
    },
  },
}
`;
